include/rmm/mr/device/hybrid_memory_resource.hpp

```diff
@@ -0,0 +1,155 @@
+   * @brief Queries whether the resource supports use of non-null streams for
+   * allocation/deallocation.
+   *
+   * @returns true
+   *---------------------------------------------------------------------------**/
+  bool supports_streams() const noexcept override { return true; }
+
+ private:
+
+  rmm::mr::device_memory_resource& get_allocator(std::size_t bytes) {
+    if (bytes > max_fixed_size_) return *large_size_mr_;
+    else {
+      auto iter = fixed_size_mr_.begin();
+      while (bytes > iter->get()->get_block_size())
+        iter = std::next(iter);
+      if (iter != fixed_size_mr_.end())
+        return *iter->get();
+      else
+        throw std::bad_alloc();
+    }
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Allocates memory of size at least \p bytes.
+   *
+   * The returned pointer will have at minimum 256 byte alignment.
+   *
+   * @throws std::bad_alloc if size > block_size (constructor parameter)
+   *
+   * @param bytes The size of the allocation
+   * @param stream Stream on which to perform allocation
+   * @return void* Pointer to the newly allocated memory
+   *---------------------------------------------------------------------------**/
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override {
+    if (bytes <= 0) return nullptr;
+    return get_allocator(bytes).allocate(bytes, stream);
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Deallocate memory pointed to by \p p.
+   *
+   * @throws std::bad_alloc if size > block_size (constructor parameter)
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes The size in bytes of the allocation. This must be equal to the
+   * value of `bytes` that was passed to the `allocate` call that returned `p`.
+   * @param stream Stream on which to perform deallocation
+   *---------------------------------------------------------------------------**/
+  void do_deallocate(void* p, std::size_t bytes, cudaStream_t stream) override {
+    get_allocator(bytes).deallocate(p, bytes, stream);
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Get free and available memory for memory resource
+   *
+   * @throws std::runtime_error if we could not get free / total memory
+   *
+   * @param stream the stream being executed on
+   * @return std::pair with available and free memory for resource
+   *---------------------------------------------------------------------------**/
+  std::pair<std::size_t, std::size_t> do_get_mem_info( cudaStream_t stream) const override {
+    return std::make_pair(0, 0);
+  }
+
+  std::size_t max_fixed_size_; // power of 2 maximum fixed size allocator
+
+  sub_memory_resource *large_size_mr_; // allocator for larger than max_fixed_size
+
+  // allocators for fixed-size blocks <= max_fixed_size_
+  std::vector<std::unique_ptr<fixed_size_memory_resource>> fixed_size_mr_;
```

Comment:
Likewise, this could probably just be a `vector<fixed_size_memory_resource>`.

Reply:
I started out with that, but for some reason the destructor was being called in the loop where I initialize them.  And see above about needing to control deallocation order since there are dependencies.

Reply:
I tried again, and could not get this to work with `vector<fixed_size_memory_resource>` because it calls the destructor for some reason as part of the emplace_back...

Reply:
Oh yeah, that makes sense. It's because when a `vector` grows, it deallocates/destroys the old contents after copying/moving them to the new allocation.